### ISSUE TITLE
ecs-agent 1.17.2 stop connection to AWS and start reporting expired credential to task

### ISSUE THREAD

**Issue by CyrilDevOps** [aws/amazon-ecs-agent#1331]
### Summary
One of our ecs-agent stop connecting to ecs and start giving expired credential to tasks running in docker

### Description
After 7 days one ecs-agnet stop connecting to ECS, and start giving expired credential
to tasks running in docker.
after speaking with aws support, we did a docker stop <ecs agent container id>
and docker start <ecs agent container id>, and the ecs-agent start working correctly again.

AWS wasn't able to explain why it stopped connecting back.

```
/var/log/ecs/ecs-agent.log.2018-04-05-01:2018-04-05T01:32:07Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-05-01:2018-04-05T01:57:38Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-05-02:2018-04-05T02:13:13Z [INFO] ACS Websocket connection closed for a valid reason
<--- docker stop docker start
/var/log/ecs/ecs-agent.log.2018-04-05-20:2018-04-05T20:06:23Z [INFO] ACS Websocket connection closed for a valid reason
```

### Expected Behavior
We want it to continue to connect to ecs and keep giving correct credential.

### Environment Details
docker 17.09
on centos 7.3.7611
Version : Amazon ECS Agent - v1.17.2 (edc3e260)

# Log (sorry it was on a prod cluster, can't get too many log)
```
/var/log/ecs/ecs-agent.log.2018-04-05-01:2018-04-05T01:32:07Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-05-01:2018-04-05T01:57:38Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-05-02:2018-04-05T02:13:13Z [INFO] ACS Websocket connection closed for a valid reason
<--- docker stop docker start
/var/log/ecs/ecs-agent.log.2018-04-05-20:2018-04-05T20:06:23Z [INFO] ACS Websocket connection closed for a valid reason
```

**Comment by aaithal** [comment]
Hi @CyrilDevOps , the ECS agent constantly disconnects and reconnects to our backend and `ACS Websocket connection closed for a valid reason` is a genuine and benign log line. We'd recommend reaching out to dev support with ECS agent log files or emailing them to aithal at amazon dot com. More instructions on gathering these logs can be found [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-logs-collector.html). Also let us know when the disconnection occurred so that we can correlate that to the logs. 

Thanks,
Anirudh

**Comment by CyrilDevOps** [comment]
We had a chat with aws support on that, and upload all the log, info they wanted (CASE 4976983061)
After studying the log, they confirm that even if ecs agent was running, it stopped to connect back to aws and an unknown reason.
They ask us to try to stop/start the agent and it seems it has started working again.
Today I had the same problem, on another node, with an uptime of 11 days, the agent stop connecting to aws this morning, and now 12 hours later,
I had to do a docker restart <ecs agent docker id>
to make it work again.

**Comment by adnxn** [comment]
@CyrilDevOps, we suspect the issue may be related to a stale websocket fix that was included with ecs-agent [v1.17.3](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.17.3). We recommend upgrading the agent to the latest version.

Otherwise the current set of agent logs do not provide sufficient visibility into the issue you are reporting. If you continue to experience this problem with the latest version of the agent, please update the issue. Additionally, it would be very helpful if you can narrow down the exact time agent stopped connecting to the ECS backend.

**Comment by adnxn** [comment]
> What sort of timing details to you need to help 'debug' the problem ?

Would it be possible to run your agent with debug level logging? You can do this by setting the `ECS_LOGLEVEL=debug` in the `ecs.config`. If we have debug level logging around the time it disconnects from ECS (which is normal behavior) but then fails to reconnect, it would give us visibility into what the agent is doing instead of reconnecting to the ECS backend.

**Comment by CyrilDevOps** [comment]
Sad news, I had another occurrence of ecs agent disconnected this weekend and when I checked in aws, it was the last version of ecs agent, so the problem isn't fixed.
/var/log/ecs/ecs-agent.log.2018-04-15-08:2018-04-15T08:51:30Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-15-08:2018-04-15T08:51:30Z [INFO] Connected to ACS endpoint
/var/log/ecs/ecs-agent.log.2018-04-15-09:2018-04-15T09:06:49Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-16-02:2018-04-16T02:40:44Z [INFO] Connected to ACS endpoint

/var/log/ecs/ecs-agent.log.2018-04-16-02:2018-04-16T02:40:44Z [INFO] Amazon ECS agent Version: 1.17.3, Commit: 159ae5c3

**Comment by CyrilDevOps** [comment]
we are doing our monthly patch rollover, I added the ECS_LOGLEVEL=debug in it, and will try to
get more info next time I see that problem.
we had that problem three of four time in production, on a cluster where nodes aren't very active, the instances uptime is 3 weeks to a month.

**Comment by CyrilDevOps** [comment]
We add the issue again this weekend, between last ecs agent and docker 17.09.1 on centos 7
the instance/docker/ecs-agent was up for 3 days.
I get all the log, collect, the stack trace ..., however as it is a production server I can share them here, maybe if we can have a corresponding aws number to do the exchange.

I guess I can share the stack trace itself.

get ACS in ecs log :
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:45:43Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:45:46Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:45:46Z [DEBUG] Received connect to ACS message
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:45:46Z [INFO] Connected to ACS endpoint
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:45:46Z [DEBUG] ACS client starting websocket poll loop
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:46:41Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:47:37Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:48:32Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:49:27Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:50:22Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:51:20Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:52:15Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:53:10Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:54:05Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:55:00Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:55:55Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:56:50Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:57:45Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:58:40Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-14:2018-04-22T14:59:35Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:00:30Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:01:25Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:02:20Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:03:15Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:04:10Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:05:05Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:06:00Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:06:55Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:07:50Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:08:45Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:09:41Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:10:36Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:11:31Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:12:26Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:13:21Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:14:16Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:15:11Z [DEBUG] ACS activity occurred
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:15:43Z [INFO] ACS Websocket connection closed for a valid reason
/var/log/ecs/ecs-agent.log.2018-04-22-15:2018-04-22T15:15:43Z [DEBUG] Received connect to ACS message
/var/log/ecs/ecs-agent.log.2018-04-23-01:github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).startACSSession(0xc420233500, 0x100d200, 0xc420223d60, 0x1011200, 0xc420306000, 0x100d300, 0xc42022c930, 0xc4201ab260, 0x100f300, 0xc4202002d0, ...)

**Comment by aaithal** [comment]
Hi @CyrilDevOps, can you please send these logs to aithal at amazon dot com?

Thanks,
Anirudh

**Comment by CyrilDevOps** [comment]
I just send you the collect.tgz.

**Comment by CyrilDevOps** [comment]
The file is too big for your email, can you enable a place where I can drop it ? or any info on what subset of file you need ?

**Comment by CyrilDevOps** [comment]
Aaithal, the account is incorrect according to AWS.

**Comment by aaithal** [comment]
It's most likely because of the s3 console. Can you try sharing it with ecs-debug-external at amazon dot com?

**Comment by aaithal** [comment]
Hi @CyrilDevOps, thank you so much sending us those logs. The ECS agent gathers information about the Docker daemon such as version information when connecting to ECS backend in order to receive container/task start/stop commands. From the stack trace you sent, it seems to be stuck waiting for a response from Docker when invoking the `Version` API (See snippet from stack trace below). 

We are not sure why Docker daemon is taking such a long time to respond to the `Version()` call (We also noticed that you might be running a custom Docker version, not sure if it's related). But, we are modifying the ECS agent to add a timeout around this call and are also modifying the ECS agent to optimize that call path a little bit more. We'll post a PR link soon for this.

If/when this happens again, please share the stack trace with us so that we can correlate these failures if possible.

Thanks,
Anirudh

```
goroutine 1 [select, 633 minutes]:
net/http.(*persistConn).roundTrip(0xc42024e000, 0xc4203eaae0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1970 +0x608
net/http.(*Transport).RoundTrip(0xc4202760f0, 0xc4207ac200, 0xc4202760f0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:413 +0x992
net/http.send(0xc4207ac200, 0x10062c0, 0xc4202760f0, 0x0, 0x0, 0x0, 0xc4201fe038, 0xbbf1e0, 0xc420755470, 0x1)
	/usr/local/go/src/net/http/client.go:249 +0x1a9
net/http.(*Client).send(0xc4201e83f0, 0xc4207ac200, 0x0, 0x0, 0x0, 0xc4201fe038, 0x0, 0x1, 0x410bf7)
	/usr/local/go/src/net/http/client.go:173 +0xfd
net/http.(*Client).Do(0xc4201e83f0, 0xc4207ac200, 0xc42001a038, 0xc4207ac200, 0xc420755768)
	/usr/local/go/src/net/http/client.go:602 +0x28d
github.com/aws/amazon-ecs-agent/agent/vendor/golang.org/x/net/context/ctxhttp.Do(0x100e0c0, 0xc42001a038, 0xc4201e83f0, 0xc4207ac000, 0xf, 0x0, 0xc4207ac000)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/golang.org/x/net/context/ctxhttp/ctxhttp.go:30 +0x6e
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient.(*Client).do(0xc420159560, 0xbfea56, 0x3, 0xc01372, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient/client.go:472 +0x30c
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient.(*Client).Version(0xc420159560, 0x0, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient/misc.go:19 +0xad
github.com/aws/amazon-ecs-agent/agent/engine.(*dockerGoClient).Version(0xc4202d6980, 0x10, 0xad4220, 0xc4200ae001, 0xc420534240)
	/go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_client.go:962 +0x59
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).Version(0xc420306000, 0xc420534240, 0xc006fc, 0x6, 0xc4200ae0e8)
	/go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_task_engine.go:1105 +0x37
github.com/aws/amazon-ecs-agent/agent/acs/handler.acsWsURL(0xc4200154a0, 0x28, 0xc42001604c, 0xe, 0xc4203e0000, 0x5a, 0x1011200, 0xc420306000, 0x7f4ef4ba8280, 0xc420430890, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:417 +0x3fa
github.com/aws/amazon-ecs-agent/agent/acs/handler.(*session).startSessionOnce(0xc420088a90, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:248 +0x279
github.com/aws/amazon-ecs-agent/agent/acs/handler.(*session).Start(0xc420088a90, 0x1, 0x1)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:195 +0x19b
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).startACSSession(0xc420233500, 0x100d200, 0xc420223d60, 0x1011200, 0xc420306000, 0x100d300, 0xc42022c930, 0xc4201ab260, 0x100f300, 0xc4202002d0, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:576 +0x1a5
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).doStart(0xc420233500, 0xc42022c700, 0x100d200, 0xc420223d60, 0x1012860, 0xc4202001e0, 0x100fcc0, 0xc420220900, 0x100f300, 0xc4202002d0, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:298 +0x790
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).start(0xc420233500, 0xc42001a038)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:201 +0x339
github.com/aws/amazon-ecs-agent/agent/app.Run(0xc42000e180, 0x0, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/run.go:64 +0x15b
main.main()
	/go/src/github.com/aws/amazon-ecs-agent/agent/agent.go:29 +0x63
```

**Comment by CyrilDevOps** [comment]
Our docker version is 17.09.1 CE on centos 7.4
We are not using the last (18.03) docker because there is some problem with
cgroup and memory leak with redhat kernel, so we are stuck with 17.09.1 until redhat fix it.
We also use aqua (aqua-agent 2.6 if I remember correctly), it filter docker command and image.
During it's installation (after the start of ecs-agent), it replace the docker unix socker with its own,
so if ecs-agent connect/disconnect regularly using this unix socket, the go command will through aqua agent first I guess.

**Comment by aaithal** [comment]
> During it's installation (after the start of ecs-agent), it replace the docker unix socker with its own,
so if ecs-agent connect/disconnect regularly using this unix socket, the go command will through aqua agent first I guess.

I'm not sure if that's what's causing this issue. But, the second stack trace you sent also points to the `docker version` API just never returning because of which ECS agent seems to be stuck as well (I have pasted the stack trace below). This should be an inexpensive API and I'm not sure why that API is taking sol or seemingly stuck. 

As per my earlier comment, we're working on adding a timeout for this call in the ECS agent as a workaround. However, the root-cause is probably something related to your Docker installation. 

```
goroutine 1 [select, 26 minutes]:
net/http.(*persistConn).roundTrip(0xc420199e60, 0xc4201d3b00, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1970 +0x608
net/http.(*Transport).RoundTrip(0xc4202100f0, 0xc420143500, 0xc4202100f0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:413 +0x992
net/http.send(0xc420143500, 0x10062c0, 0xc4202100f0, 0x0, 0x0, 0x0, 0xc42000c1d0, 0xbbf1e0, 0xc420541470, 0x1)
	/usr/local/go/src/net/http/client.go:249 +0x1a9
net/http.(*Client).send(0xc4201d22d0, 0xc420143500, 0x0, 0x0, 0x0, 0xc42000c1d0, 0x0, 0x1, 0x410bf7)
	/usr/local/go/src/net/http/client.go:173 +0xfd
net/http.(*Client).Do(0xc4201d22d0, 0xc420143500, 0xc42001a028, 0xc420143500, 0xc420541768)
	/usr/local/go/src/net/http/client.go:602 +0x28d
github.com/aws/amazon-ecs-agent/agent/vendor/golang.org/x/net/context/ctxhttp.Do(0x100e0c0, 0xc42001a028, 0xc4201d22d0, 0xc420143400, 0xf, 0x0, 0xc420143400)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/golang.org/x/net/context/ctxhttp/ctxhttp.go:30 +0x6e
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient.(*Client).do(0xc420200090, 0xbfea56, 0x3, 0xc01372, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient/client.go:472 +0x30c
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient.(*Client).Version(0xc420200090, 0x0, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/fsouza/go-dockerclient/misc.go:19 +0xad
github.com/aws/amazon-ecs-agent/agent/engine.(*dockerGoClient).Version(0xc420259680, 0x10, 0xad4220, 0xc420095b01, 0xc4204ca780)
	/go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_client.go:962 +0x59
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).Version(0xc42029c000, 0xc4204ca780, 0xc006fc, 0x6, 0xc420095c68)
	/go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_task_engine.go:1105 +0x37
github.com/aws/amazon-ecs-agent/agent/acs/handler.acsWsURL(0xc4200142a0, 0x28, 0xc42001604c, 0xe, 0xc420338780, 0x5a, 0x1011200, 0xc42029c000, 0x7f3cf69e9338, 0xc4204aa540, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:417 +0x3fa
github.com/aws/amazon-ecs-agent/agent/acs/handler.(*session).startSessionOnce(0xc42006c5b0, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:248 +0x279
github.com/aws/amazon-ecs-agent/agent/acs/handler.(*session).Start(0xc42006c5b0, 0x1, 0x1)
	/go/src/github.com/aws/amazon-ecs-agent/agent/acs/handler/acs_handler.go:195 +0x19b
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).startACSSession(0xc4201f5a00, 0x100d200, 0xc42026c8c0, 0x1011200, 0xc42029c000, 0x100d300, 0xc4202127e0, 0xc420212070, 0x100f300, 0xc420216190, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:576 +0x1a5
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).doStart(0xc4201f5a00, 0xc4202125b0, 0x100d200, 0xc42026c8c0, 0x1012860, 0xc4202160a0, 0x100fcc0, 0xc420201440, 0x100f300, 0xc420216190, ...)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:298 +0x790
github.com/aws/amazon-ecs-agent/agent/app.(*ecsAgent).start(0xc4201f5a00, 0xc42001a028)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/agent.go:201 +0x339
github.com/aws/amazon-ecs-agent/agent/app.Run(0xc42000e180, 0x0, 0x0, 0x0)
	/go/src/github.com/aws/amazon-ecs-agent/agent/app/run.go:64 +0x15b
main.main()
	/go/src/github.com/aws/amazon-ecs-agent/agent/agent.go:29 +0x63
```

**Comment by petderek** [comment]
This was released in 1.18.0.
